pkgs/development/compilers/cudatoolkit/flags.nix

```diff
@@ -136,10 +142,15 @@ let
         base = gencodeMapper "sm" cudaCapabilities;
         forward = gencodeMapper "compute" [ (lists.last cudaCapabilities) ];
       in
-      base ++ lib.optionals enableForwardCompat forward;
+      base ++ lists.optionals enableForwardCompat forward;
   };
 
 in
+# All of the requested config.cudaCapabilities *must* be in supportedCapabilities.
+assert asserts.assertMsg
+  (unsupportedCapabilities == [ ])
+  ("CUDA version ${cudaVersion} does not support the following requested capabilities: "
+    + strings.concatStringsSep ", " unsupportedCapabilities);
```

Comment:
There's been recently a discussion about moving asserts into `meta.broken` in the pytorch expression, and I'm now unsure if we should keep introducing new asserts into `cudaPackages`. The two concerns are:

1. The user cannot disable the assert in their overlay (fixed by introducing some `config.cudaAllowUnsupportedCapabilities`)
2. We could get a good error message displayed in the wrong place

But looking at your `nix build .#magma` example, I actually like the output: the distance between "CUDA does not support ..." and "evaluating attribute 'cmakeFlags' of derivation 'magma-2.7.1'" is just a few lines. That's pretty easy to read?

What happens if you try to `nix build .#python3Packages.torch` which consumes the broken `magma`?

Reply:
Is there any mechanism to provide a message along with `meta.broken` status? Or perhaps an option to provide a warning instead of an assert? Minimizing the number of asserts is a noble goal IMHO -- In this case I suppose there are UX tradeoffs either way

Reply:
@SomeoneSerge building `torch` hides the source of the error :(

```console
$ nix build --impure -L .#python3Packages.torch
error:
       … while evaluating the attribute 'drvPath'

         at /nix/store/9pm8mkf5dp11715mfn865xxd26snz8n9-source/lib/customisation.nix:228:7:

          227|     in commonAttrs // {
          228|       drvPath = assert condition; drv.drvPath;
             |       ^
          229|       outPath = assert condition; drv.outPath;

       … while evaluating the attribute 'drvPath'

         at /nix/store/9pm8mkf5dp11715mfn865xxd26snz8n9-source/lib/customisation.nix:228:7:

          227|     in commonAttrs // {
          228|       drvPath = assert condition; drv.drvPath;
             |       ^
          229|       outPath = assert condition; drv.outPath;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: CUDA version 11.7 does not support the following requested capabilities: 8.9
```

Reply:
> Is there any mechanism to provide a message along with `meta.broken` status? Or perhaps an option to provide a warning instead of an assert?

There is, I think -- we could compute a term `isBroken` which we pass to `meta.broken`, but wrap it in one of the [`lib.warnIf` family](https://teu5us.github.io/nix-lib.html#lib.trivial.warnif). That would allow us to warn the user that this is unsupported, but would not break the build with an assert (rather it would be broken by way of being marked as broken, and have a message printed).

I'll try to refactor to do that as an example -- the goal being to mark a version of cudaPackages as broken for a certain configuration if there are no valid requested capabilities.

Reply:
Interesting! I don't see any other packages doing that, and I'm not 100% certain if this is consistent with the "no warnings from top-level attributes" policy